tag/template_service.go

```diff
@@ -0,0 +1,158 @@
+package tag
+
+import (
+	"fmt"
+
+	"github.com/odpf/columbus/tag/validator"
+	"github.com/pkg/errors"
+)
+
+var validDomainFieldDataType = []string{
+	"string",
+	"double",
+	"boolean",
+	"enumerated",
+	"datetime",
+}
+
+// Service is a type of service that manages business process
+type TemplateService struct {
+	validator  validator.Validator
+	repository TemplateRepository
+}
+
+// Validate validates domain template based on the business rule
+func (s *TemplateService) Validate(template Template) error {
+	err := s.validator.Validate(template)
+	if err != nil {
+		err = ValidationError{err}
+	}
+
+	return err
+}
+
+// Create handles create business operation for template
+func (s *TemplateService) Create(template *Template) error {
+	if template == nil {
+		return errors.New("template is nil")
+	}
+	err := s.Validate(*template)
+	if err != nil {
+		return err
+	}
+
+	filterForExistence := Template{
+		URN: template.URN,
+	}
+	templateRecords, err := s.repository.Read(filterForExistence)
+	if err != nil {
+		return errors.Wrap(err, "error checking template existence")
+	}
+	if len(templateRecords) > 0 {
+		return DuplicateTemplateError{URN: template.URN}
+	}
+
+	err = s.repository.Create(template)
+	if err != nil {
+		return errors.Wrap(err, "error creating template")
+	}
+
+	return nil
+}
+
+// Index handles read business operation for template
+func (s *TemplateService) Index(template Template) ([]Template, error) {
+	output, err := s.repository.Read(template)
+	if err != nil {
+		return nil, errors.Wrap(err, "error fetching templates")
+	}
+	return output, nil
+}
+
+// Update handles update business operation for template
+func (s *TemplateService) Update(template *Template) error {
+	if template == nil {
+		return errors.New("template is nil")
+	}
+	err := s.Validate(*template)
+	if err != nil {
+		return err
+	}
+	filterForExistence := Template{
+		URN: template.URN,
+	}
+	templateRecords, err := s.repository.Read(filterForExistence)
+	if err != nil {
+		return errors.Wrap(err, "error checking template existence")
+	}
+	if len(templateRecords) == 0 {
```

Comment:
I guess this should be done in Read repository layer

Reply:
Read currently returns list of templates, and it is also being used for fetching list of templates by TemplateService.